Add capabilities support

gdm85 opened the pull request:
I have now completed the PR, it allows adding/dropping capabilities through [libcap-ng](https://people.redhat.com/sgrubb/libcap-ng/) :)

The added CLI option is `--caps` and takes a comma-separated argument list like `-all,+chown,+setuid`. The first capability of the list can be an alias like `+all`/`all`, `-all` (the `+` sign can be omitted); I loosely modeled it on this one: docker/docker#6687.

Examples of supported syntax:

```
pflask --caps +all # this would be the default, so it respects current version's behavior
pflask --caps -all,+chown # drop all except chown
pflask --caps -all,chown # same as above
pflask --caps +all,-mknod,-setgid,-setuid # allow all capabilities except mknod,setgid and setuid
pflask --caps all,-mknod,-setgid,-setuid # same as above
pflask --caps -mknod,-setgid,-setuid # same as above
```

Example of usage:

```
# touch test && pflask --caps -all chown 1000.1000 test
chown: changing ownership of ‘test’: Operation not permitted
[✘] Child failed with code '1'
# touch test && pflask --caps -all,+chown chown 1000.1000 test
[✔] Child exited
```

gdm85 committed:
Add capabilities support (with libcap-ng)

Case-insensitive compare for the 'all' alias too, to match libcap-ng functionality

Do not build capabilities without libcap-ng

gdm85 commented:
I like how caps were proposed initially for Docker here: https://github.com/docker/docker/issues/6687

I will add some commits on this PR to implement it for pflask like this:
```
pflask --cap +ALL # this would be the default, so it respects current version's behavior
pflask --cap -ALL,+CHOWN # drop all except chown
pflask --cap +ALL,-MKNOD,-SETGID,-SETUID # allow all except mknod,setgid and setuid
```

If you have a better idea, please tell :)

gdm85 committed:
Add --caps option through gengetopt

gdm85 commented:
I've added what I think is a simple syntax for the capabilities feature.
```
# touch test && pflask --cap -all chown 1000.1000 test
chown: changing ownership of ‘test’: Operation not permitted
[✘] Child failed with code '1'
# touch test && pflask --cap -all,+chown chown 1000.1000 test
[✔] Child exited
```

gdm85 committed:
Version bump to 0.3

gdm85 commented:
The last commit is about increasing the version, as I thought this would be a notable change.

Now that I think of it, man page update is missing..and maybe something else?

gdm85 committed:
Update documentation about --caps option

ghedo commented on src/pflask.c:
```diff
@@ -44,6 +44,11 @@
 #include <sys/wait.h>
 
+#if HAVE_LIBCAP_NG
+#include <cap-ng.h>
```
Is this include actually needed? Also, the "capabilities.h" can be left outside of the `#if`.

gdm85 replied:
I've made `capabilities.c` (in `wscript`) built only when libcap-ng is available, hence the conditional `capabilities.h` include. But yes, I've removed the cap-ng.h include

ghedo commented on src/capabilities.c:
```diff
@@ -0,0 +1,110 @@
+#include <syslog.h>
+
+#include <pwd.h>
+#include <grp.h>
```
Many of the includes above aren't actually needed.

gdm85 replied:
You're right, removed them.

ghedo commented on src/capabilities.c:
```diff
@@ -0,0 +1,110 @@
+	int rc;
+	capng_act_t cap_action;
+	char *cap, *name;
+	char *remainder = ",";
```
This should be `:`, to be more consistent with other options.

ghedo replied:
Nevermind, "," is ok.

ghedo commented on src/capabilities.c:
```diff
@@ -0,0 +1,110 @@
+	capng_get_caps_process();
+
+	// get first cap
+	cap = strtok(caps, remainder);
```
Instead of using `strtok()` directly, the code should be refactored to use `split_str()` from util.h, making it somewhat easier to follow.

ghedo replied:
Actually. a better idea would be to add the `multiple` flag to --caps and then process the list directly. See how this is done for e.g. --setenv.

gdm85 replied:
Re-implemented using `multiple`

ghedo commented:
I added some comments inline, thanks for your work!

gdm85 committed:
Use multiple arguments list feature of gengetopt
Move --caps option to top (alphabetical sorting)

gdm85 commented on src/cmdline.ggo:
```diff
@@ -27,6 +27,8 @@ option "setenv"    s "Set additional environment variables"
        string optional multiple
 option "keepenv"   k "Do not clear environment"
        flag off
+option "caps"  b "Specify comma-separated capabilities to add ('+' prefix) or drop ('-' prefix)"
```
Note how I moved this to top, since I noticed there is alphabetical order

gdm85 commented:
@ghedo thanks to you for pflask :)

I processed your feedback, the build is successful but please note it's not really building the code for this new feature as it's an optional library and it's not available there for Travis

gdm85 commented on src/cmdline.ggo:
```diff
@@ -1,6 +1,8 @@
+version "0.3"
 
+option "caps"  b "Specify comma-separated capabilities to add ('+' prefix) or drop ('-' prefix)"
+       string default="+all" optional multiple
```
Only drawback I see with doing it this way is that if there are multiple ``all`` specified, and they get de-duplicated, then you get an error about ``all`` not being the first argument. E.g.:
```
# pflask --caps all,-chown,all
[✘] Alias 'all' is valid only as first capability
[✘] Child failed with code '1'
```

It's not documented and might leave the user surprised, but I think it's okay to leave it like this.

gdm85 commented on src/capabilities.c:
```diff
@@ -0,0 +1,104 @@
+
+		rc = capng_name_to_capability(name);
+		if (rc == -1) {
+			fail_printf("Invalid capability name: '%s'", name);
```
Note that user will get this error also when they're specifying only prefixes, e.g. `--caps +,-,-`, although the "empty capability name" error (as per line 49) could have been a better match.

ghedo commented on src/capabilities.c:
```diff
@@ -0,0 +1,104 @@
+	for (unsigned int i = 0; i < caps_given; i++) {
+		cap = caps[i];
+
+		sys_fail_if(strlen(cap) == 0, "Empty capability name specified");
```
A few things. The `sys_fail_if()` is only intended to be used when you expect `errno` to be set (internally `strerror()` gets called and the error returned is printed. This is not the case here.

Anyway, the option validation should probably happen in `main()` before any work is actually done so that we print an error as soon as possible (same goes for the all check with i != 0, and maybe you could add the + and - check as well).

ghedo commented on src/capabilities.c:
```diff
@@ -0,0 +1,104 @@
+	}
+
+	rc = capng_apply(CAPNG_SELECT_BOTH);
+	if (rc != 0) {
```
This would be a good place to use the `fail_if()` macro. Same for the two rc checks above as well.

gdm85 committed:
Replaced some invalid sys_fail_if() calls (ghedo)

gdm85 committed:
Perform validation in main() (ghedo)
Introduce struct cap_action to store parsed valid capability actions
Introduce memory_cleanup() function to release buffers at exit

gdm85 committed:
Silence some warnings

gdm85 commented:
@ghedo I've processed your feedback, however I am not particularly fond of the introduced `memory_cleanup()` to prevent leakage (in case of errors), and of the conditional `#if`s, if you have better ideas let me know.

The memory cleanup (called with `atexit()` is kind of necessary because there is not a precise execution flow control e.g. pflask can exit to operative system from many places in the code, and the allocated buffer for capability actions would leak memory.

gdm85 commented on src/capabilities.c:
```diff
@@ -0,0 +1,55 @@
+		fail_if(rc != 0, "Error updating capabilities");
+	}
+
+	rc = capng_apply(CAPNG_SELECT_BOTH);
```
perhaps this could be skipped in case of `clear_caps == false && total_caps == 0`, but it shouldn't do any harm

gdm85 commented:
Alternatively, one could allow only `-all,+cap1,+cap2...` or `+all,-cap1,-cap2...` semantics and since we know the list of cap names has been de-duplicated, we could skip memory allocations and use a fixed-size array by using `CAP_LAST_CAP` (http://lxr.free-electrons.com/source/include/uapi/linux/capability.h#L355) as upper bound; in such case however specifying something like `-all,+cap1,-cap2` would trigger an error, albeit it would be perfectly legit to request such calls on both the libcap-ng and kernel subsystems. I have not proposed this approach because if only those two semantics are allowed, then we could rather go for two mutually-exclusive command-line options, `--caps-drop` and `--caps-allow`.

ghedo commented:
You are making it way more complicated than it needs to be :)

I pushed 7dc26dc with several changes. Can you please check that I didn't break anything?

Thanks again for your work!

ghedo commented:
I also configured Travis CI to install dependencies, so this is now covered too.

gdm85 commented:
@ghedo thanks, that's cool! I've added some comments on 7dc26dc8bc1c600d06eb1a0ecc2e329171dedc53 functionally it's fine, I made some tests. I assume you don't bother freeing buffers as they are released anyway on process exit.

Perhaps the home page README needs a section too for this feature?

By the way, clever code you added there, I am definitively rusty in C :smile:
